File: server/api_offer.py
from flask import Blueprint, request, jsonify, abort
from datetime import date
from flask import current_app
import os
import jwt
from functools import wraps
import logging
from db_offer import *
from create import session
from db_auth import *
from create import *
from config import *
from setup import *
from security import *


offer = Blueprint("offer", __name__)
"""
api offer
"""
@offer.route("/api/v1/offer/all_all", methods=["GET"])
@token_required
def get_all_offers(current_user):
    try:
        offers = session.query(Offer).all()
        responce = []
        for offer in offers:
            responce.append({
                "id":offer.id,
                "tool_name": offer.tool_name ,
                "tool_description": offer.tool_description ,
                "location": offer.location ,
                "price": offer.price ,
                "date_start": offer.date_start ,
                "date_finish": offer.date_finish,
                "owner_name": offer.owner_name,
                "phone_number": offer.phone_number
            })
        return responce
    except Exception as e:
        current_app.logger.info("exception: ", e)
        # return jsonify(500, f"Error occured in {e}")


@offer.route("/api/v1/offer/save", methods=["POST"])
@token_required
def save_offer(current_user):
    try:
        if not request.json:
            abort(400)
        token = request.json["Authorization"]
        data = request.json
        user_info = jwt.decode(token, SECRET_KEY, algorithms=['HS256'])["username"]
        user = session.query(User).filter(User.username == user_info).first()
        user_offers = user.offers
        offer = Offer(
            data["tool_name"], data["tool_description"], 
            data["location"], data["price"], data["date_start"],
            data["date_finish"], data["owner_name"], data["phone_number"]
        )
        user_offers.append(offer)
        session.add(offer)
        session.commit()
        return {"status": "success"}
    except Exception as e:
        current_app.logger.info("%s failed to log in", user.username)
        # return jsonify(500, f"Error occured in {e}")


@offer.route("/api/v1/offer/all", methods=["GET"])
@token_required
def get_all_user_offers(current_user):
    try:
        token = request.headers["Authorization"]
        user_info = jwt.decode(token, SECRET_KEY, algorithms=["HS256"])["username"]
        user = session.query(User).filter(User.username == user_info).first()
        user_offers = user.offers
        responce  = []
        for offer in user_offers:
            responce.append({
                "id": offer.id,
                "tool_name": offer.tool_name,
                "tool_description": offer.tool_description,
                "location": offer.location,
                "price": offer.price,
                "date_start": offer.date_start,
                "date_finish": offer.date_finish,
                "owner_name": offer.owner_name,
                "phone_number": offer.phone_number
            })
        return {"result": responce}
    except Exception as e:
        current_app.logger.info("%s failed to log in", user.username)
        # return jsonify(500, f"Error occured in {e}")

File: client/lib/api/ApiOffer.dart
import 'dart:convert';

import 'package:client/api/ApiMethodsImpl.dart';
import 'package:client/api/utils.dart';
import '../models/offer.dart';

class ApiOffer extends APIMethodsImpl {
  // 1) add an offer to user
  // url: "/api/v1/offer/save
  // method: POST
  Future<int> registerOfferToUser(toolName, toolDescription, price,
      dateStart, dateFinish, lat, lng, ownerName, phoneNumber) async {
    String url = "$URL/api/v1/offer/save";
    Map<String, String> credits = {
      "tool_name": toolName,
      "tool_description": toolDescription,
      "price": price,
      "date_start": dateStart,
      "date_finish": dateFinish,
      "lat": lat,
      "lng": lng,
      "owner_name": ownerName,
      "phone_number": phoneNumber,
    };
    final response = await APIMethodsImpl().post(url, credits);
    var status = response.statusCode;
    return status;
  }
  //  2) get all offers
  // url: "/api/v1/offer/all_all"
  // method: GET
  Future<List<Offer>> fetchOffers() async {
    String url = "$URL/api/v1/offer/all_all";
    final response = await APIMethodsImpl().get(url);
    if (response.statusCode == 200) {
      // print(json.decode(response.body)["data"]);
      final List<dynamic> jsonList = json.decode(response.body)["data"];
      // print(jsonList.map((json) => Offer.fromJson(json)).toList());
      return jsonList.map((json) => Offer.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load posts');
    }
  }
  // 3) get all offer of a user
  // URL: ”/api/v1/offer/all”
  // method: GET
  Future<List<Offer>> fetchUserOffers() async {
    String url = "$URL/api/v1/offer/all";
    final response = await APIMethodsImpl().get(url);
    if (response.statusCode == 200) {
      final List<dynamic> jsonList = json.decode(response.body)["data"];
      return jsonList.map((json) => Offer.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load posts');
    }
  }
  // 4) delete offer by id
  // URL: “/api/v1/offer/delete“
  // method: DELETE
  Future<int> deleteOffer(id) async {
    String url = "$URL/api/v1/offer/delete";
    Map body = {"id": id};
    final response = await APIMethodsImpl().delete(url, body,);
    return response.statusCode;
  }
  // 5) get offer by query
  // URL: “/api/v1/offer/query“
  // method: POST
  Future<List<Offer>> queryOffer(query) async {
    String url = "$URL/api/v1/offer/query";
    Map credits = {
      "query": query,
    };
    final response = await APIMethodsImpl().post(url, credits);
    if (response.statusCode == 200) {
      final List<dynamic> jsonList = json.decode(response.body)["data"];
      return jsonList.map((json) => Offer.fromJson(json)).toList();
    } else {
      throw Exception('Failed to load offers');
    }
  }
  // 6) update particular offer
  // URL: “/api/v1/offer/upd“
  // method: PUT
}